// @salesforce/commerce-sdk-react/hooks/ShopperProducts/queryKeyHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProducts = exports.getProduct = exports.getCategory = exports.getCategories = void 0;
var _utils = require("../utils");
/*
 * Copyright (c) 2023, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

// We must use a client with no parameters in order to have required/optional match the API spec

// This is defined here, rather than `types.ts`, because it relies on `Client` and `QueryKeys`,
// and making those generic would add too much complexity.

const getProducts = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'ids', 'inventoryIds', 'currency', 'expand', 'locale', 'allImages', 'perPricebook', 'siteId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/products'],
  queryKey: params => [...getProducts.path(params), getProducts.parameters(params)]
};
exports.getProducts = getProducts;
const getProduct = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'id', 'inventoryIds', 'currency', 'expand', 'locale', 'allImages', 'perPricebook', 'siteId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/products/', params.id],
  queryKey: params => [...getProduct.path(params), getProduct.parameters(params)]
};
exports.getProduct = getProduct;
const getCategories = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'ids', 'levels', 'locale', 'siteId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/categories'],
  queryKey: params => [...getCategories.path(params), getCategories.parameters(params)]
};
exports.getCategories = getCategories;
const getCategory = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'id', 'levels', 'locale', 'siteId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/categories/', params.id],
  queryKey: params => [...getCategory.path(params), getCategory.parameters(params)]
};
exports.getCategory = getCategory;
