// @salesforce/pwa-kit-runtime/ssr/server/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.X_ORIGINAL_CONTENT_TYPE = exports.X_MOBIFY_QUERYSTRING = exports.X_MOBIFY_FROM_CACHE = exports.STRICT_TRANSPORT_SECURITY = exports.STATIC_ASSETS = exports.SET_COOKIE = exports.PROXY_PATH_PREFIX = exports.NO_CACHE = exports.CONTENT_TYPE = exports.CONTENT_SECURITY_POLICY = exports.CONTENT_ENCODING = exports.CACHE_CONTROL = exports.BUILD = exports.APPLICATION_OCTET_STREAM = void 0;
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const APPLICATION_OCTET_STREAM = 'application/octet-stream';
exports.APPLICATION_OCTET_STREAM = APPLICATION_OCTET_STREAM;
const BUILD = 'build';
exports.BUILD = BUILD;
const STATIC_ASSETS = 'static_assets';
exports.STATIC_ASSETS = STATIC_ASSETS;
const PROXY_PATH_PREFIX = '/mobify/proxy';
// All these values MUST be lower case
exports.PROXY_PATH_PREFIX = PROXY_PATH_PREFIX;
const CONTENT_TYPE = 'content-type';
exports.CONTENT_TYPE = CONTENT_TYPE;
const CONTENT_ENCODING = 'content-encoding';
exports.CONTENT_ENCODING = CONTENT_ENCODING;
const X_ORIGINAL_CONTENT_TYPE = 'x-original-content-type';
exports.X_ORIGINAL_CONTENT_TYPE = X_ORIGINAL_CONTENT_TYPE;
const X_MOBIFY_QUERYSTRING = 'x-mobify-querystring';
exports.X_MOBIFY_QUERYSTRING = X_MOBIFY_QUERYSTRING;
const X_MOBIFY_FROM_CACHE = 'x-mobify-from-cache';
exports.X_MOBIFY_FROM_CACHE = X_MOBIFY_FROM_CACHE;
const SET_COOKIE = 'set-cookie';
exports.SET_COOKIE = SET_COOKIE;
const CACHE_CONTROL = 'cache-control';
exports.CACHE_CONTROL = CACHE_CONTROL;
const NO_CACHE = 'max-age=0, nocache, nostore, must-revalidate';
exports.NO_CACHE = NO_CACHE;
const CONTENT_SECURITY_POLICY = 'content-security-policy';
exports.CONTENT_SECURITY_POLICY = CONTENT_SECURITY_POLICY;
const STRICT_TRANSPORT_SECURITY = 'strict-transport-security';
exports.STRICT_TRANSPORT_SECURITY = STRICT_TRANSPORT_SECURITY;
