// @salesforce/commerce-sdk-react/hooks/ShopperLogin/queryKeyHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getWellknownOpenidConfiguration = exports.getUserInfo = exports.getJwksUri = void 0;
var _utils = require("../utils");
/*
 * Copyright (c) 2023, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

// We must use a client with no parameters in order to have required/optional match the API spec

// This is defined here, rather than `types.ts`, because it relies on `Client` and `QueryKeys`,
// and making those generic would add too much complexity.

const getUserInfo = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'channel_id']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/oauth2/userinfo'],
  queryKey: params => [...getUserInfo.path(params), getUserInfo.parameters(params)]
};
exports.getUserInfo = getUserInfo;
const getWellknownOpenidConfiguration = {
  parameters: params => (0, _utils.pick)(params, ['organizationId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/oauth2/.well-known/openid-configuration'],
  queryKey: params => [...getWellknownOpenidConfiguration.path(params), getWellknownOpenidConfiguration.parameters(params)]
};
exports.getWellknownOpenidConfiguration = getWellknownOpenidConfiguration;
const getJwksUri = {
  parameters: params => (0, _utils.pick)(params, ['organizationId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/oauth2/jwks'],
  queryKey: params => [...getJwksUri.path(params), getJwksUri.parameters(params)]
};
exports.getJwksUri = getJwksUri;
