// @salesforce/commerce-sdk-react/hooks/ShopperOrders/queryKeyHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTaxesFromOrder = exports.getPaymentMethodsForOrder = exports.getOrder = void 0;
var _utils = require("../utils");
/*
 * Copyright (c) 2023, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

// We must use a client with no parameters in order to have required/optional match the API spec

// This is defined here, rather than `types.ts`, because it relies on `Client` and `QueryKeys`,
// and making those generic would add too much complexity.

const getOrder = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'orderNo', 'siteId', 'locale']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/orders/', params.orderNo],
  queryKey: params => [...getOrder.path(params), getOrder.parameters(params)]
};
exports.getOrder = getOrder;
const getPaymentMethodsForOrder = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'orderNo', 'siteId', 'locale']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/orders/', params.orderNo, '/payment-methods'],
  queryKey: params => [...getPaymentMethodsForOrder.path(params), getPaymentMethodsForOrder.parameters(params)]
};
exports.getPaymentMethodsForOrder = getPaymentMethodsForOrder;
const getTaxesFromOrder = {
  parameters: params => (0, _utils.pick)(params, ['organizationId', 'orderNo', 'siteId']),
  path: params => ['/commerce-sdk-react', '/organizations/', params.organizationId, '/orders/', params.orderNo, '/taxes'],
  queryKey: params => [...getTaxesFromOrder.path(params), getTaxesFromOrder.parameters(params)]
};
exports.getTaxesFromOrder = getTaxesFromOrder;
