// @salesforce/commerce-sdk-react/hooks/ShopperBaskets/cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cacheUpdateMatrix = void 0;
var _queryKeyHelpers = require("./queryKeyHelpers");
var _queryKeyHelpers2 = require("../ShopperCustomers/queryKeyHelpers");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); } /*
                                                                                                                                                                                                                                                                                                                                                                                           * Copyright (c) 2023, Salesforce, Inc.
                                                                                                                                                                                                                                                                                                                                                                                           * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                           * SPDX-License-Identifier: BSD-3-Clause
                                                                                                                                                                                                                                                                                                                                                                                           * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
                                                                                                                                                                                                                                                                                                                                                                                           */
/** Data returned by every Shopper Baskets endpoint (except `deleteBasket`) */

/** Data returned by `getCustomerBaskets` */

/** Parameters that get passed around, includes client config and possible parameters from other endpoints */

/** Parameters that we actually send to the API for `getCustomerBaskets` */

const invalidateCustomerBasketsQuery = (customerId, parameters) => {
  return {
    queryKey: _queryKeyHelpers2.getCustomerBaskets.queryKey(_objectSpread(_objectSpread({}, parameters), {}, {
      customerId
    }))
  };
};
const updateCustomerBasketsQuery = (customerId, parameters, response) => {
  return {
    queryKey: _queryKeyHelpers2.getCustomerBaskets.queryKey(_objectSpread(_objectSpread({}, parameters), {}, {
      customerId
    })),
    updater: oldData => {
      var _oldData$baskets;
      if (!(oldData !== null && oldData !== void 0 && (_oldData$baskets = oldData.baskets) !== null && _oldData$baskets !== void 0 && _oldData$baskets.length)) {
        return {
          baskets: [response],
          total: 1
        };
      }
      const updatedBaskets = oldData.baskets.map(basket => basket.basketId === parameters.basketId ? response : basket);
      return _objectSpread(_objectSpread({}, oldData), {}, {
        // Shopper Customers and Shopper Baskets have different definitions for the `Basket`
        // type. (99% similar, but that's not good enough for TypeScript.)
        // TODO: Remove this type assertion when the RAML specs match.
        baskets: updatedBaskets
      });
    }
  };
};
const cacheUpdateMatrix = {
  addCouponToBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  addGiftCertificateItemToBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  addItemToBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  addPaymentInstrumentToBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  addPriceBooksToBasket(customerId, {
    parameters
  }) {
    return {
      invalidate: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, {
        queryKey: _queryKeyHelpers.getPriceBooksForBasket.queryKey(parameters)
      },
      // TODO: Convert invalidate to an update that removes the matching basket
      ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])]
    };
  },
  addTaxesForBasket(customerId, {
    parameters
  }) {
    return {
      invalidate: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, {
        queryKey: _queryKeyHelpers.getTaxesFromBasket.queryKey(parameters)
      },
      // TODO: Convert invalidate to an update that removes the matching basket
      ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])]
    };
  },
  addTaxesForBasketItem(customerId, {
    parameters
  }) {
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }]
    };
  },
  createBasket(customerId, {
    parameters
  }, response) {
    const {
      basketId
    } = response;
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [...(customerId && !basketId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(_objectSpread(_objectSpread({}, parameters), {}, {
          basketId
        }))
      }, ...(customerId && basketId ? [updateCustomerBasketsQuery(customerId, _objectSpread(_objectSpread({}, parameters), {}, {
        basketId
      }), response)] : [])]
    };
  },
  createShipmentForBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  deleteBasket(customerId, {
    parameters
  }) {
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      remove: [
      // We want to fuzzy match all queryKeys with `basketId` in their path
      // [`/commerce-sdk-react,/organizations/,${organization},/baskets/,${basketId}`]
      {
        queryKey: _queryKeyHelpers.getBasket.path(parameters)
      }]
    };
  },
  mergeBasket(customerId, {
    parameters
  }, response) {
    const {
      basketId
    } = response;
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [...(customerId && !basketId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(_objectSpread(_objectSpread({}, parameters), {}, {
          basketId
        }))
      }, ...(customerId && basketId ? [updateCustomerBasketsQuery(customerId, _objectSpread(_objectSpread({}, parameters), {}, {
        basketId
      }), response)] : [])]
    };
  },
  removeCouponFromBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  removeGiftCertificateItemFromBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  removeItemFromBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  removePaymentInstrumentFromBasket(customerId, {
    parameters
  }, response) {
    return {
      invalidate: [{
        queryKey: _queryKeyHelpers.getPaymentMethodsForBasket.queryKey(parameters)
      },
      // TODO: Convert invalidate to an update that removes the matching basket
      ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  removeShipmentFromBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  transferBasket(customerId, {
    parameters
  }, response) {
    const {
      basketId
    } = response;
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [...(customerId && !basketId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(_objectSpread(_objectSpread({}, parameters), {}, {
          basketId
        }))
      }, ...(customerId && basketId ? [updateCustomerBasketsQuery(customerId, _objectSpread(_objectSpread({}, parameters), {}, {
        basketId
      }), response)] : [])]
    };
  },
  updateBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateBillingAddressForBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateCustomerForBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateGiftCertificateItemInBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateItemInBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updatePaymentInstrumentInBasket(customerId, {
    parameters
  }, response) {
    return {
      invalidate: [{
        queryKey: _queryKeyHelpers.getPaymentMethodsForBasket.queryKey(parameters)
      },
      // TODO: Convert invalidate to an update that removes the matching basket
      ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateShipmentForBasket(customerId, {
    parameters
  }, response) {
    return {
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateShippingAddressForShipment(customerId, {
    parameters
  }, response) {
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [{
        queryKey: _queryKeyHelpers.getShippingMethodsForShipment.queryKey(parameters)
      }, ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  },
  updateShippingMethodForShipment(customerId, {
    parameters
  }, response) {
    return {
      // TODO: Convert invalidate to an update that removes the matching basket
      invalidate: [{
        queryKey: _queryKeyHelpers.getShippingMethodsForShipment.queryKey(parameters)
      }, ...(customerId ? [invalidateCustomerBasketsQuery(customerId, parameters)] : [])],
      update: [{
        queryKey: _queryKeyHelpers.getBasket.queryKey(parameters)
      }, ...(customerId ? [updateCustomerBasketsQuery(customerId, parameters, response)] : [])]
    };
  }
};
exports.cacheUpdateMatrix = cacheUpdateMatrix;
